File: frontend/src/stores/instruments.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/services/api'

// Per-point dollar multipliers, keyed by symbol -- fetched once from the
// backend's single source of truth (app/core/instruments.py) rather than
// hard-coded here, per VS3's rule against scattering instrument values
// through Vue components.
export const useInstrumentsStore = defineStore('instruments', () => {
  const multipliers = ref({})

  async function fetchMultipliers() {
    const { data } = await api.get('/instruments/multipliers')
    const parsed = {}
    for (const [symbol, value] of Object.entries(data)) {
      parsed[symbol] = Number(value)
    }
    multipliers.value = parsed
    return parsed
  }

  // Mirrors backend/app/core/instruments.py's _DATED_CONTRACT_RE: a dated
  // futures contract like "MNQZ26" is root "MNQ" + single-letter month code
  // + 1-4 digit year. Only this parsing pattern is duplicated here -- the
  // multiplier values themselves still come from the fetched table above,
  // the one source of truth for those.
  const DATED_CONTRACT_RE = /^([A-Z]{1,3})[FGHJKMNQUVXZ]\d{1,4}$/

  function getMultiplier(symbol) {
    if (!symbol) return null
    const normalized = symbol.trim().toUpperCase()
    const known = multipliers.value[normalized]
    if (known !== undefined) return known

    const match = normalized.match(DATED_CONTRACT_RE)
    if (match) {
      const rootKnown = multipliers.value[match[1]]
      if (rootKnown !== undefined) return rootKnown
    }
    return null
  }

  return { multipliers, fetchMultipliers, getMultiplier }
})

File: backend/app/api/v1/instruments.py
"""Instrument reference data (VS3): exposes the centralized multiplier
table (app.core.instruments) so the frontend's risk/reward preview reads
from the single source of truth instead of hard-coding per-instrument
dollar values in Vue components."""
from fastapi import APIRouter, Depends

from app.api.deps import get_current_user
from app.core.instruments import INSTRUMENT_MULTIPLIERS
from app.models.user import User

router = APIRouter()


@router.get("/instruments/multipliers")
async def get_instrument_multipliers(current_user: User = Depends(get_current_user)) -> dict[str, str]:
    return {symbol: str(value) for symbol, value in INSTRUMENT_MULTIPLIERS.items()}
